langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdin/StdioMcpTransport.java

```diff
@@ -0,0 +1,132 @@
+    private Process process;
+    private ProcessIOHandler processIOHandler;
+    private final Map<Long, CompletableFuture<JsonNode>> pendingOperations = new ConcurrentHashMap<>();
+    private final boolean logEvents;
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final Logger log = LoggerFactory.getLogger(StdioMcpTransport.class);
+
+    public StdioMcpTransport(Builder builder) {
+        this.command = builder.command;
+        this.environment = builder.environment;
+        this.logEvents = builder.logEvents;
+    }
+
+    @Override
+    public void start() {
+        log.info("Starting process: {}", command);
+        ProcessBuilder processBuilder = new ProcessBuilder(command);
+        processBuilder.environment().putAll(environment);
+        try {
+            process = processBuilder.start();
+            log.debug("PID of the started process: {}", process.pid());
+            process.onExit().thenRun(() -> {
+                log.debug("Subprocess has exited with code: {}", process.exitValue());
+            });
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        processIOHandler = new ProcessIOHandler(process, pendingOperations, logEvents);
+        // FIXME: where should we obtain the thread?
+        new Thread(processIOHandler).start();
+        new Thread(new ProcessStderrHandler(process)).start();
+    }
+
+    @Override
+    public JsonNode initialize(McpInitializeRequest request) {
+        try {
+            String requestString = OBJECT_MAPPER.writeValueAsString(request);
+            return executeAndWait(requestString, request.getId());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public JsonNode listTools(McpListToolsRequest operation) {
+        try {
+            String requestString = OBJECT_MAPPER.writeValueAsString(operation);
+            return executeAndWait(requestString, operation.getId());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public JsonNode executeTool(McpCallToolRequest operation) {
+        try {
+            String requestString = OBJECT_MAPPER.writeValueAsString(operation);
+            return executeAndWait(requestString, operation.getId());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        process.destroy();
+    }
+
+    private JsonNode executeAndWait(String request, Long id) {
+        try {
+            CompletableFuture<JsonNode> future = new CompletableFuture<>();
+            pendingOperations.put(id, future);
+            processIOHandler.submit(request);
+            return future.get();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```

Comment:
Wouldn't it be good to log something in case of error?

Reply:
This exception will bubble up to the application into the `DefaultMcpClient` constructor, so I'd say it's not necessary?!

Reply:
Depending on if we want to see something in the log from StdioMcpTransport. I'm fine to skip it